staging/src/k8s.io/mount-utils/mount_linux_test.go

```diff
@@ -654,7 +654,7 @@ func TestFormat(t *testing.T) {
 		formatCount    = 5
 		fstype         = "ext4"
 		output         = "complete"
-		cmdDuration    = 1 * time.Millisecond
+		cmdDuration    = 5 * time.Millisecond
```

Comment:
the problem is that the test is relying on small timers, once a cmd finish it removes on execution.

This test logic is very racy, if we want to test we achieved the expected concurrency we should start creating goroutines and the "fake command" should wait until we achieve the number of gorutiness expected, and timeout if they are not achieved.
Then trying to create a new command and see how it never runs and it gets blocked until one of the commands expire

Reply:
@TommyStarK can you think in other way of testing it?
I feel something is odd here with this implementation or with the test, I don't know what

Reply:
I have the same feeling, especially those [lines](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/mount-utils/mount_linux_test.go#L702-L713). Yes I will think about testing in a different way.

Reply:
@aojea did you have something like that in mind (or I am completely missing the point):

```go
func TestFormat(t *testing.T) {
	const (
		formatCount = 5
		fstype      = "ext4"
		output      = "complete"
		cmdDuration = 1 * time.Millisecond
	)

	tests := []struct {
		desc    string
		max     int
		timeout time.Duration
	}{
		{
			max:     2,
			timeout: 1 * time.Nanosecond,
		},
		{
			max:     3,
			timeout: 10 * time.Nanosecond,
		},
		{
			max:     4,
			timeout: 100 * time.Nanosecond,
		},
	}

	for _, tc := range tests {
		t.Run(fmt.Sprintf("max=%d,timeout=%s", tc.max, tc.timeout.String()), func(t *testing.T) {
			var mu sync.Mutex
			exec := &testexec.FakeExec{}
			maxConcurrent := tc.max
			starter := make(chan struct{})
			witness := make(chan struct{})

			for i := 0; i < formatCount; i++ {
				exec.CommandScript = append(exec.CommandScript, makeFakeCommandAction(output, nil, func() {
					time.Sleep(cmdDuration)
				}))
			}
			mounter := NewSafeFormatAndMount(nil, exec, WithMaxConcurrentFormat(maxConcurrent, tc.timeout))

			var wg sync.WaitGroup
			for i := 0; i < maxConcurrent+1; i++ {
				c := i
				wg.Add(1)
				go func() {
					<-starter
					defer wg.Done()
					mounter.format(fstype, nil) // should block once maxConcurrent is reached

					mu.Lock()
					if c == maxConcurrent {
						close(witness)
					}
					mu.Unlock()
				}()
			}

			close(starter)

			select {
			case <-witness:
				t.Fatal("max concurrency should have been reached")
			case <-time.After(tc.timeout):
				return
			}

			wg.Wait()
		})
	}
}
```

Reply:
I was thinking on blocking on the command execution, forget about the cmdDuration for a moment.

You set concurrency to N
The command execution increments a counter and blocks on a channel waiting.
You execute N+1 command in parallel, the counter has to be N
Send N struct{}{} through the channel, the counter has to be 1

Reply:
I like the test, but let's split the test in two `TestFormatConcurrency` and `TestFormatTimeout`.

`TestFormatConcurrency` is the one you are showing, just add a close to the channel and verify that the concurrency is 0.

`TestFormatTimeout` is another similar test, just that we don't need permutations, block on max concurrency, execute a new one that has to be blocked for the timeout period, and then concurrency will go to n+1 (I think this is a bug but this is the implemented behavior, IMHO this should have been implemented passing a context to the Exec command with the timeout https://pkg.go.dev/os/exec#CommandContext , and using golang rate-limiter https://pkg.go.dev/golang.org/x/time/rate

Reply:
Sure thing, will do as soon as I got the time this week. Thanks 😊

Reply:
@aojea I've pushed a commit including `TestFormatConcurrency` with your feedback and `TestFormatTimeout` following your scenario (hope I understood it properly). I'm still wondering why the implemented behaviour allows to go to N+1 concurrency when I blocked on the configured max concurrency and execute a new `format` and wait for the timeout period. Tbh I am not sure to have implemented and understood exactly what you asked for, so, as usual, if there is anything please correct me and I'll do the changes. Thank you for your time.

Reply:
> Tbh I am not sure to have implemented

based on the discussion on the original PR it seems they just wanted an easy escape hatch

> >  IMHO this should have been implemented passing a context to the Exec command with the timeout https://pkg.go.dev/os/exec#CommandContext , and using golang rate-limiter https://pkg.go.dev/golang.org/x/time/rate

you can implement it properly instead of just fixing only the tests ^^^

or fix the tests and then improve it, up to you

Reply:
- stress `TestFormatConcurrency`
<img width="388" alt="Screenshot 2023-05-04 at 18 44 00" src="https://user-images.githubusercontent.com/9576370/236271585-9fc9dd0e-722e-4293-8449-29a8e4d338f2.png">

- stress `TestFormatTimeout`
<img width="315" alt="Screenshot 2023-05-04 at 18 44 06" src="https://user-images.githubusercontent.com/9576370/236271643-7fbf8d9e-2cee-4203-b481-2e306564f819.png">

If you don't mind, I'd rather keep this PR to fix the flaky test and open a fresh new one where I could implement the command context and the rate limiter. As usual, thanks for everything Antonio.